### PR TITLE
agents, concepts: better changed api

### PR DESCRIPTION
**Pull request by rerowep**
* Adds `resolve` to POST parameter.
* Without `resolve` only `pid`, `_created`, `_updated` and `deleted`` will be returned.
* Are unknown pids will be reported back without additional information. Example `{'pid', '42'}

Co-Authored-by: Peter Weber <peter.weber@rero.ch>

### CODE REVIEW

File: .github/workflows/continuous-integration-test.yml [hunk 1]
```diff
@@ -37,14 +37,14 @@ jobs:
...
 
-    - name: Use Cache
-      uses: actions/cache@v3
-      with:
-        path: |
-          ~/.cache/pip
-          ~/.cache/pypoetry/virtualenvs
-          ~/.npm
-        key: ${{ runner.os }}-pip-venv-${{ hashFiles('**/poetry.lock') }}
+    # - name: Use Cache
```
Comment (jma) [inline .github/workflows/continuous-integration-test.yml:40]: To be removed?

File: pyproject.toml [hunk 2]
```diff
@@ -74,6 +74,8 @@ Werkzeug = "<2.0.0"
 click = "<8.0.0"
 sentry-sdk = "<1.6.1"
 pytest-invenio = ">=1.4.0,<1.4.12"
+jsonref = "<1.0.0"
```
Comment (jma) [inline pyproject.toml:77]: Did you use `poetry add` to avoid all update dependencies?

File: rero_mef/agents/mef/api.py [hunk 3]
```diff
@@ -217,12 +217,29 @@ def get_updated(cls, data):
...
         if pids := data.get('pids'):
+            # find deleted pids
+            all_pids = [hit.pid for hit in search.source('pid').scan()]
+            for missing_pid in sorted(list(set(pids) - set(all_pids))):
+                data = {'pid': missing_pid}
+                mef = AgentMefRecord.get_record_by_pid(
+                    missing_pid,
+                    with_deleted=True
+                )
+                if mef == {}:
```
Comment (jma) [inline rero_mef/agents/mef/api.py:235]: `if not mef`? Or better `if mef.is_deleted` -> https://github.com/inveniosoftware/invenio-records/blob/ce7989d84e418c9aed7d4e0bfcaffba8a10f0253/invenio_records/api.py#L105

File: rero_mef/agents/mef/api.py [hunk 4]
```diff
@@ -217,12 +217,29 @@ def get_updated(cls, data):
...
+                    data['_created'] = mef.created.isoformat()
```
Comment (jma) [inline rero_mef/agents/mef/api.py:236]: Do we have to take timezone for this?

File: rero_mef/concepts/mef/api.py [hunk 5]
```diff
@@ -162,15 +162,29 @@ def get_updated(cls, data):
         :param pid: pid to use..
         :returns: latest record.
         """
-        search = ConceptMefSearch()
+        search = ConceptMefSearch()\
```
Comment (jma) [inline rero_mef/concepts/mef/api.py:165]: This code seems very similar to the Agent class. Can it be in a parent class?

File: rero_mef/utils.py [hunk 6]
```diff
@@ -1243,11 +1243,18 @@ def get_mefs_endpoints():
...
     """Lagging genarator."""
     yield '['
-    for idx, record in enumerate(search.scan()):
+    idx = 0
+    for hit in search.scan():
         if idx != 0:
             yield ', '
-        yield json.dumps(record.to_dict())
+        record = hit.to_dict()
+        yield json.dumps(record)
```
Comment (jma) [inline rero_mef/utils.py:1254]: Why not `yield json.dumps(hit.to_dict())`?

File: tests/api/test_agents_mef_rest.py [hunk 7]
```diff
@@ -145,6 +145,11 @@ def test_agents_mef_get_updated(client, agent_mef_record, agent_idref_record,
                                 agent_mef_idref_redirect_record,
                                 agent_mef_gnd_redirect_record):
     """Test agents MEF get latest."""
+
+    def utf_isoformat(date):
+        """Changed datetime to UTC isoformaz."""
```
Comment (jma) [inline tests/api/test_agents_mef_rest.py:150]: isoformaz -> isoformat

File: tests/api/test_agents_mef_rest.py [hunk 8]
```diff
@@ -145,6 +145,11 @@ def test_agents_mef_get_updated(client, agent_mef_record, agent_idref_record,
                                 agent_mef_idref_redirect_record,
                                 agent_mef_gnd_redirect_record):
     """Test agents MEF get latest."""
+
+    def utf_isoformat(date):
```
Comment (jma) [inline tests/api/test_agents_mef_rest.py:149]: Are you sure that no lib can do this?

File: tests/api/test_agents_mef_rest.py [hunk 9]
```diff
@@ -153,6 +158,26 @@ def test_agents_mef_get_updated(client, agent_mef_record, agent_idref_record,
...
+    }, {
+        'pid': agent_mef_idref_redirect_record.pid,
+        '_created': utf_isoformat(agent_mef_idref_redirect_record.created),
+        '_updated': utf_isoformat(agent_mef_idref_redirect_record.updated)
+    }]
+
+    res, data = postdata(
+        client,
+        'api_blueprint.agent_mef_get_updated',
+        {'resolve': 1}
```
Comment (jma) [inline tests/api/test_agents_mef_rest.py:178]: It seems that you do not test that the response contains the resolved values.

**Review by jma — CHANGES_REQUESTED** [review]

File: pyproject.toml (hunk 2 above)
Comment (rerowep) [reply pyproject.toml:77]: No I wanted to update the other packages too.

File: rero_mef/agents/mef/api.py (hunk 3 above)
Comment (rerowep) [reply rero_mef/agents/mef/api.py:235]: Does not give the same result!

File: rero_mef/utils.py (hunk 6 above)
Comment (rerowep) [reply rero_mef/utils.py:1254]: ok

File: tests/api/test_agents_mef_rest.py (hunk 8 above)
Comment (rerowep) [reply tests/api/test_agents_mef_rest.py:149]: This is the simplest solution for ES DB datetimes.

File: tests/api/test_agents_mef_rest.py (hunk 9 above)
Comment (rerowep) [reply tests/api/test_agents_mef_rest.py:178]: Added

File: rero_mef/api_mef.py [hunk 10]
```diff
@@ -215,6 +214,42 @@ def mark_as_deleted(self, dbcommit=False, reindex=False):
...
+            .params(preserve_order=True) \
+            .sort({'pid': {'order': 'asc'}})
+        deleted = []
+        if from_date := data.get('from_date'):
+            search = search.filter('range', _updated={'gte': from_date})
+        if pids := data.get('pids'):
+            # find deleted pids
+            missing_pids = []
+            for pid in pids:
+                if cls.search().filter('term', pid=pid).count() == 0:
```
Comment (jma) [inline rero_mef/api_mef.py:235]: Can probably be optimized.

File: rero_mef/api_mef.py [hunk 11]
```diff
@@ -215,6 +214,42 @@ def mark_as_deleted(self, dbcommit=False, reindex=False):
...
+                    missing_pids.append(pid)
+            for missing_pid in missing_pids:
+                data = {'pid': missing_pid}
+                mef = cls.get_record_by_pid(
```
Comment (jma) [inline rero_mef/api_mef.py:239]: You can get using id instead of pid for performance purpose.

File: tests/api/test_agents_mef_rest.py [hunk 12]
```diff
@@ -153,26 +160,67 @@ def test_agents_mef_get_updated(client, agent_mef_record, agent_idref_record,
...
+    )
+    assert res.status_code == 200
     pids = sorted([rec.get('pid') for rec in data])
     assert pids == ['1', '2', '3']
```
Comment (jma) [inline tests/api/test_agents_mef_rest.py:184]: Seems not robust.

File: tests/api/test_agents_mef_rest.py [hunk 13]
```diff
@@ -181,5 +229,58 @@ def test_agents_mef_get_updated(client, agent_mef_record, agent_idref_record,
...
     assert res.status_code == 200
-    pids = sorted([rec.get('pid') for rec in data])
-    assert pids == []
+    assert data == []
+
+    date = datetime.now(timezone.utc) + timedelta(days=1)
+    res, data = postdata(
+        client,
+        'api_blueprint.agent_mef_get_updated',
+        {"from_date": date.isoformat(), "pids": ['2', '4']}
```
Comment (jma) [inline tests/api/test_agents_mef_rest.py:238]: Seems not robust. Do not use contant values if it is possible.

**Review by jma — APPROVED** [review]

**Pull request merged**